cava/CoqStyleGuide.md

```diff
@@ -0,0 +1,254 @@
+     highlight them separately.
+- One `Require Import` statement per line; it's easier to scan that way.
+- `Require Import` statements should use "fully-qualified" names (e.g. `Require
+  Import Coq.ZArith.ZArith` instead of `Require Import ZArith`).
+  * Use the `Locate` command to find the fully-qualified name!
+- `Require Import`s should go in the following order:
+   1. Standard library dependencies (start with `Coq.`)
+   2. External dependencies (anything outside the current project)
+   3. Same-project dependencies
+- `Require Import`s with the same root library (the name before the first `.`)
+should be grouped together. Within each root-library group, they should be in
+alphabetical order (so `Coq.Lists.List` before `Coq.ZArith.ZArith`).
+
+### Notations and scopes
+
+- Any file-wide `Local Open Scope`s should come immediately after the `Import`s
+  (see example).
+  * Always use `Local Open Scope`; just `Open Scope` will sneakily
+open the scope for those who import your file.
+- Put notations in their own separate modules or files, so that those who
+  import your file can choose whether or not they want the notations.
+  * Conflicting notations can cause a lot of headache, so it comes in very
+    handy to leave this flexibility!
+
+## Formatting
+
+### Line length
+
+- Maximum line length 80 characters.
+  * Many Coq IDE setups divide the screen in half vertically and use only half
+    to display source code, so more than 80 characters can be genuinely hard to
+    read on a laptop.
+
+### Whitespace and indentation
+
+- No trailing whitespace.
+- Spaces, not tabs.
+- Files should end with a newline.
+  * Many editors do this automatically on save.
+- Default indentation is 2 spaces.
+  * Keeping this small prevents complex proofs from being indented ridiculously
+    far, and matches IDE defaults.
+- Use 2-space indents if inserting a line break immediately after:
+  * `Proof.`
+  * `fun <...> =>`
+  * `forall <...>,`
+  * `exists <....>,`
+- The style for indenting arguments in function application depends on where
+  you make a line break.  If you make the line break immediately after the
+function name, use a 2-space indent.  However, if you make it after one or more
+arguments, align the next line with the first argument:
+  ```coq
+  (Z.pow
+     1 2)
+  (Z.pow 1 2 3
+         4 5 6)
+  ```
+- `Inductive` cases should not be indented. Example:
+  ```coq
+  Inductive Foo : Type :=
+  | FooA : Foo
+  | FooB : Foo
+  .
+  ```
+- `match` or `lazymatch` cases should line up with the "m" in `match` or "l" in
+  `lazymatch`, as in the following examples:
+  ```coq
+  match x with
+  | 3 => true
+  | _ => false
+  end.
+
+  lazymatch x with
+  | 3 => idtac
+  | _ => fail "Not equal to 3:" x
+  end.
+
+  repeat match goal with
+         | _ => progress subst
+         | _ => reflexivity
+         end.
+
+  do 2 lazymatch goal with
+       | |- context [eq] => idtac
+       end.
+  ```
```

Comment:
is there an advantage to keeping them aligned with the match/lazymatch vs unindented? e.g.

```
 repeat match goal with
 | _ => progress subst
 | _ => reflexivity
 end.
```

I don't feel strongly either way, but having it aligned to the outer scope might require less formatting?

Reply:
The reason I added it is that the emacs + ProofGeneral setup that I use automatically aligns them with the `match`, so I have to constantly manually re-indent if the match is aligned differently. I thought coqide indented it the same way -- out of curiosity, what setup are you using?

If I were to back-solve for a more philosophical explanation, there's kind of an invisible parenthesis around the `match...end`, so it makes some sense to format a `repeat match` similarly to a function application.

Reply:
> emacs + ProofGeneral setup that I use automatically aligns

Thats good enough reason for me! 

> what setup are you using?

I'm using vscode, with as many changes/plugins to make it as close to vim as possible :) (Not using vim directly as I had issues with existing Coq vim plugins)